### PR TITLE
Enable CA2208: Instantiate argument exceptions correctly

### LINKED ISSUE #13909: Fix CA2208: Instantiate argument exceptions correctly
_Tracking issue._

 Various [CA2208: Instantiate argument exceptions correctly](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2208) rule violations are suppressed in #13898.

### PR DESCRIPTION
**Pull request by xtqqczze**
Fix: https://github.com/PowerShell/PowerShell/issues/13909.

https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2208

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs [hunk 1]
```diff
@@ -579,7 +579,7 @@ private void ProcessMTUSize(string targetNameOrAddress)
                 WriteObject(new PingMtuStatus(
                     Source,
                     resolvedTargetName,
-                    replyResult ?? throw new ArgumentNullException(nameof(replyResult)),
+                    replyResult ?? throw new SystemException(nameof(replyResult)),
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs:582]: I believe we shouldn't use SystemException because it is _generic base_ class.

Comment (xtqqczze) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs:582]: I can refactor this code to not throw, in new PR.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs [hunk 2]
```diff
@@ -985,8 +985,8 @@ public class PingMtuStatus : PingStatus
             /// <param name="destination">The destination machine name of the ping.</param>
             /// <param name="reply">The response from the ping attempt.</param>
             /// <param name="bufferSize">The buffer size from the successful ping attempt.</param>
-            internal PingMtuStatus(string source, string destination, PingReply reply, int bufferSize)
-                : base(source, destination, reply, 1)
+            internal PingMtuStatus(string source, string destination, PingReply? reply, int bufferSize)
+                : base(source, destination, reply ?? throw new SystemException(nameof(reply)), 1)
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs:989]: @xtqqczze Why do you mess style and functional changes in one PR?
If you think we should change exception types you should do this in separate PR. (Although I see no point in it). Please fix only exception arguments in the PR.

Comment (xtqqczze) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs]: Changes to exception types are required to enable CA2208, if `ArgumentException` is the wrong type. The exception type must change if the `paramName` parameter cannot be made to match the name of a parameter.

**Review by iSazonov — COMMENTED** [review]
Please revert types. We shouldn't change types.

**Comment by xtqqczze** [comment]
@iSazonov I will revert all changes not part of the CodeFix, and suppress the rule with the justification linking to the tracking issue: #13909

File: src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs [hunk 3]
```diff
@@ -479,6 +479,10 @@ private void ProcessTraceroute(string targetNameOrAddress)
         #endregion TracerouteTest
 
         #region MTUSizeTest
+        [System.Diagnostics.CodeAnalysis.SuppressMessage(
+            "Usage",
+            "CA2208:Instantiate argument exceptions correctly",
+            Justification = "https://github.com/PowerShell/PowerShell/issues/13909")]
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs:485]: The suppression attribute is too broad.
Please use local suppression with directive and c# comment with tracking issue.

Comment (xtqqczze) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs:485]: I would prefer to avoid using preprocessor directives. Is this blocking the PR?

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs:485]: Why do you prefer to avoid this?

The attribute will suppress the rule for all current scope but there may be (in the future) other cases that fall under this rule. I think suppression should be more precise - just what is needed.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs (hunk 1 above)
Comment (xtqqczze) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs]: `SystemException` is actually a reserved exception type, using it would be a violation of [CA2201](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/CA2201)

**Comment by xtqqczze** [comment]
closing/reopening to restart tests

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConsoleColorCmdlet.cs [hunk 4]
```diff
@@ -89,6 +89,10 @@ public ConsoleColorCmdlet()
         }
 
         #region helper
+        [System.Diagnostics.CodeAnalysis.SuppressMessage(
+            "Usage",
+            "CA2208:Instantiate argument exceptions correctly",
+            Justification = "https://github.com/PowerShell/PowerShell/issues/13909")]
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConsoleColorCmdlet.cs:95]: Please use local suppression. It is difficult to find that we suppress.

Comment (xtqqczze) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConsoleColorCmdlet.cs:95]: `BuildOutOfRangeErrorRecord` is a three line method, why do we need local suppression here?

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConsoleColorCmdlet.cs:95]: There are other cases below.

**Comment by xtqqczze** [comment]
Rebased and fixed up to address @iSazonov review.

**Review by iSazonov — APPROVED** [review]

**Comment by xtqqczze** [comment]
_Blocked by #15604._

File: src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs [hunk 5]
```diff
@@ -1808,7 +1808,9 @@ private List<CompletionResult> GetResultForIdentifier(CompletionContext completi
                             case UsingStatementKind.Type:
                                 break;
                             default:
+#pragma warning disable CA2208 // Instantiate argument exceptions correctly
                                 throw new ArgumentOutOfRangeException("UsingStatementKind");
+#pragma warning restore CA2208 // https://github.com/PowerShell/PowerShell/issues/13909
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs:1813]: Why not use nameof?

Comment (xtqqczze) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs:1813]: I decided to suppress because `UsingStatementKind` is not the name of a method parameter, therefore this would arguably be incorrect use.

**Comment by iSazonov** [comment]
I think now it is more useful benefit from new APIs like `ArgumentNullException.ThrowIfNull()`. They exclude most of  CA2208.
